packages/api/lambdas/sf-event-sqs-to-db-records/index.js

```diff
@@ -56,49 +66,91 @@ const writeRecordsToDynamoDb = async (cumulusMessage) => {
+  pdrModel,
+}) => {
+  if (!isPostRDSDeploymentExecution(cumulusMessage)) {
+    return writeRecordsToDynamoDb({
+      cumulusMessage,
+      granuleModel,
+      executionModel,
+      pdrModel,
+    });
+  }
 
-  const collectionCumulusId = await getMessageCollectionCumulusId(cumulusMessage, knex);
-  const isExecutionRDSWriteEnabled = await shouldWriteExecutionToRDS(
-    cumulusMessage,
+  const messageCollectionNameVersion = getCollectionNameAndVersionFromMessage(cumulusMessage);
+  const messageAsyncOperationId = getMessageAsyncOperationId(cumulusMessage);
+  const messageParentExecutionArn = getMessageExecutionParentArn(cumulusMessage);
+  const [
     collectionCumulusId,
-    knex
-  );
+    asyncOperationCumulusId,
+    parentExecutionCumulusId,
+  ] = await Promise.all([
+    getCollectionCumulusId(
+      messageCollectionNameVersion,
+      knex
+    ),
+    getAsyncOperationCumulusId(
+      messageAsyncOperationId,
+      knex
+    ),
+    getParentExecutionCumulusId(
+      messageParentExecutionArn,
+      knex
+    ),
+  ]);
 
-  // If execution is not written to RDS, then PDRs/granules which reference
-  // execution should not be written to RDS either
-  if (!isExecutionRDSWriteEnabled) {
-    return writeRecordsToDynamoDb(cumulusMessage);
+  if (!shouldWriteExecutionToPostgres({
+    messageCollectionNameVersion,
+    collectionCumulusId,
+    messageAsyncOperationId,
+    asyncOperationCumulusId,
+    messageParentExecutionArn,
+    parentExecutionCumulusId,
+  })) {
+    // If any requirements for writing executions to Postgres were not met,
+    // then PDR/granules should not be written to Postgres either since they
+    // reference executions, so bail out to writing execution/PDR/granule
+    // records to Dynamo.
+    return writeRecordsToDynamoDb({
+      cumulusMessage,
+      granuleModel,
+      executionModel,
+      pdrModel,
+    });
   }
 
+  const executionCumulusId = await writeExecution({
+    cumulusMessage,
+    collectionCumulusId,
+    asyncOperationCumulusId,
+    parentExecutionCumulusId,
+    knex,
+  });
+
   const providerCumulusId = await getMessageProviderCumulusId(cumulusMessage, knex);
 
-  try {
-    const executionCumulusId = await writeExecution({
-      cumulusMessage,
-      knex,
-    });
-    // PDR write only attempted if execution saved
-    const pdrCumulusId = await writePdr({
-      cumulusMessage,
-      collectionCumulusId,
-      providerCumulusId,
-      knex,
-      executionCumulusId,
-    });
-    return await writeGranules({
-      cumulusMessage,
-      collectionCumulusId,
-      providerCumulusId,
-      executionCumulusId,
-      pdrCumulusId,
-      knex,
-    });
-  } catch (error) {
-    log.error(`Failed to write records for ${executionArn}`, error);
-    throw error;
-  }
+  const pdrCumulusId = await writePdr({
+    cumulusMessage,
+    collectionCumulusId,
+    providerCumulusId,
+    knex,
+    executionCumulusId,
+  });
+
+  return writeGranules({
```

Comment:
Why is awaiting no longer needed for `writeGranules`?

Reply:
Very good question. Originally, `writeGranules` the final statement in a `try` block and its value was supposed to be returned if it did not throw an error. The "return await" is really necessary for the second part - to ensure any error from the function is thrown into the `catch` block, otherwise you're just returning the un-evaluated promise of `writeGranules` so any error from it will only be thrown on a call to the parent function, `writeRecords`.  

But since `writeGranules` is no longer in a try/catch block (because I was just re-throwing the error so the try/catch wasn't really providing any value), I can just `return writeGranules` so that `await writeRecords` effectively awaits the promise returned by `writeGranules`. 

Jake Archibald explains it better than I would: https://jakearchibald.com/2017/await-vs-return-vs-return-await/